### ISSUE TITLE
VoiceOver: cannot enter text with Braille screen input

### ISSUE THREAD

**Issue by codeofdusk** [holzschu/a-shell#75]
Thanks so much for adding VoiceOver support to the app! The autoread works great!

However, I cannot enter text using VoiceOver's [Braille screen input](https://support.apple.com/guide/iphone/type-onscreen-braille-using-voiceover-iph10366cc30/ios) feature.

**Comment by holzschu** [comment]
Thanks for signaling this issue. I am going to have a look.

**Comment by codeofdusk** [comment]
Not sure how helpful this is, but a similar issue was fixed in ish-app/ish@ca909d9.

**Comment by codeofdusk** [comment]
Based on that commit, it seems you need to implement `UITextInput`.

**Comment by holzschu** [comment]
I tried, and it doesn't work. I keep getting "not recognized", even though the letters briefly appears on screen. It is also what I get if I don't implement `UITextInput`.

I might still make it work, but it will take some time.

**Comment by holzschu** [comment]
Hi,
in the latest TestFlight version, we moved to a different rendering engine (SwiftTerm), which has better support for VoiceOver, and should support Braille input as well. If you're willing to test, it available here: 
https://testflight.apple.com/join/WUdKe3f4

And I'm in the right moment to take suggestions or complaints about how the new engine integrates with VoiceOver.

**Comment by codeofdusk** [comment]
@holzschu I can confirm that this bug is fixed!

One small note though: I'm not sure how you're detecting inserted text. Since it's a terminal, you're going to have a tough time (I imagine, I did a lot of work on a major Windows screen reader's terminal support) detecting user inserted text from application inserted text. But a small thing you can do (which will have some false positives admittedly) is to suppress text insersions of one character from being reported, as this should get rid of *most* typed character insersions (not necessarily the case in Braille, as in some languages like English, Braille characters [do not map one-to-one to print ones](https://en.wikipedia.org/wiki/English_Braille#Contractions)).

**Comment by holzschu** [comment]
If you're talking about how the current version is speaking each character as it is inserted on the command line, yes, it was very annoying and I have removed that for the next version (not yet released, not even on TestFlight). I just kept reporting autocomplete options (as it is necessary for autocomplete to work) and the character under the cursor when moving with arrows.

But this is an area that is entirely customizable, so I'm taking suggestions as to what should the terminal speak when the user is writing a line on the command line.

**Comment by codeofdusk** [comment]
@holzschu If possible, the terminal should honour VoiceOver's typing echo setting and not insert arbitrary output for terminal echo. But I realize that's a hard problem to solve and you're probably going to have to rely on heuristics to filter input out

**Comment by codeofdusk** [comment]
How does your character suppression logic work? It breaks in readline-type environments. STR:

1. Open AShell.
2. At the prompt, type `python3`. Observe that only VoiceOver echo reports the entered characters (good).
3. At the Python prompt, type `import this`. Observe that characters are once again doubled (VoiceOver echo + terminal echo; bad).

**Comment by holzschu** [comment]
I see what you mean. There isn't any character suppression logic yet (which answers your question). I'll put one specifically in place for interactive commands like Python (but not for commands like vim or less).

**Comment by codeofdusk** [comment]
> There isn't any character suppression logic yet

Sorry, what I mean by that is: how are you determining what's echo-back of user input vs. output of a command run by the user? Presumably the shell just gives you back an output stream

**Comment by holzschu** [comment]
Thanks for the clarification; your description of the issue was very clear; I could connect it with the idea of "character suppression". It's currently not done, or not yet.

a-Shell has a separation between "interactive commands" and "non-interactive commands". The latter (for example nslookup or dash) let a-Shell handle the input, which in turn lets VoiceOver handle the input, so there's no local echo and no need for character suppression. The former are handling the input, so they are in charge of echoing it to the command line, hence the local echo for VoiceOver. Interactive commands include vim, less and python. The local echo issue that you pointed out will be a major issue with python, but not with less and vim. Then again, suppressing output of local commands will not hurt these either.
